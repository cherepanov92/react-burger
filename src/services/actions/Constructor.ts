import { OrderedIngredient } from '../../utils/types';

export const ADD_INGREDIENT: 'ADD_INGREDIENT' = 'ADD_INGREDIENT';
export const REMOVE_INGREDIENT: 'REMOVE_INGREDIENT' = 'REMOVE_INGREDIENT';
export const MOVE_INGREDIENT: 'MOVE_INGREDIENT' = 'MOVE_INGREDIENT';

interface IAddIngredient {
    readonly type: typeof ADD_INGREDIENT;
    ingredient: OrderedIngredient;
}

interface IRemoveIngredient {
    readonly type: typeof REMOVE_INGREDIENT;
    ingredient: OrderedIngredient;
}

interface IMoveIngredient {
    readonly type: typeof MOVE_INGREDIENT;
    ingredient: OrderedIngredient;
    newIndex: number;
}

export type TIngredientActions = IAddIngredient | IRemoveIngredient | IMoveIngredient;