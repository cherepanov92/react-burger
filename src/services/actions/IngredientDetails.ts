import { IngredientType } from '../../utils/types';

export const ADD_INGREDIENT_DETAILS: 'ADD_INGREDIENT_DETAILS' = 'ADD_INGREDIENT_DETAILS';
export const REMOVE_INGREDIENT_DETAILS: 'REMOVE_INGREDIENT_DETAILS' = 'REMOVE_INGREDIENT_DETAILS';

interface IAddIngredientDetails {
    readonly type: typeof ADD_INGREDIENT_DETAILS;
    ingredient: IngredientType;
}

interface IRemoveIngredientDetails {
    readonly type: typeof REMOVE_INGREDIENT_DETAILS;
}

export type TIngredientDetailsActions = IAddIngredientDetails | IRemoveIngredientDetails;