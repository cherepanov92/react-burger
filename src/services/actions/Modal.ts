import { EnumModalType } from '../../utils/types';

export const APPEND_MODAL_TYPE: 'APPEND_MODAL_TYPE' = 'APPEND_MODAL_TYPE';
export const APPEND_ERROR_MODAL_TYPE: 'APPEND_ERROR_MODAL_TYPE' = 'APPEND_ERROR_MODAL_TYPE';
export const REMOVE_MODAL_TYPE: 'REMOVE_MODAL_TYPE' = 'REMOVE_MODAL_TYPE';

export interface IAppendModalType {
    readonly type: typeof APPEND_MODAL_TYPE;
    modalType: EnumModalType;
}

interface IAppendErrorModalType {
    readonly type: typeof APPEND_ERROR_MODAL_TYPE;
    status?: string | null;
    message: string | null;
}

interface IRemoveModalTypeE {
    readonly type: typeof REMOVE_MODAL_TYPE;
}

export type TModalActions = IAppendModalType | IRemoveModalTypeE | IAppendErrorModalType;