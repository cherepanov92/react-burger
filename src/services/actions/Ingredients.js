import { getIngredients } from "../../utils/api";
import { APPEND_ERROR_MODAL_TYPE } from "./Modal";

export const GET_INGREDIENTS_REQUEST = 'GET_INGREDIENTS_REQUEST';
export const GET_INGREDIENTS_SUCCESS = 'GET_INGREDIENTS_SUCCESS';
export const GET_INGREDIENTS_FAILED = 'GET_INGREDIENTS_FAILED';

export const getIngredientsData = () => {
    return function(dispatch) {
        dispatch({
            type: GET_INGREDIENTS_REQUEST
        });
        getIngredients()
        .then(res => {
            if (res && res.success) {
                dispatch({
                    type: GET_INGREDIENTS_SUCCESS,
                    items: res.data
                });
            } else {
                dispatch({
                    type: GET_INGREDIENTS_FAILED
                });
            }
        })
        .catch(err => {
            dispatch({
                type: APPEND_ERROR_MODAL_TYPE,
                errorStatus: err.status
            })
        })
    };
}