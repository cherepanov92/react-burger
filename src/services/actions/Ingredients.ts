import { getIngredients } from '../../utils/api';
import { APPEND_ERROR_MODAL_TYPE } from './Modal';
import { IError, IngredientType } from '../../utils/types';
import { Dispatch } from 'react';

export const GET_INGREDIENTS_REQUEST: 'GET_INGREDIENTS_REQUEST' = 'GET_INGREDIENTS_REQUEST';
export const GET_INGREDIENTS_SUCCESS: 'GET_INGREDIENTS_SUCCESS' = 'GET_INGREDIENTS_SUCCESS';
export const GET_INGREDIENTS_FAILED: 'GET_INGREDIENTS_FAILED' = 'GET_INGREDIENTS_FAILED';

interface IGetIngredientsRequest {
    readonly type: typeof GET_INGREDIENTS_REQUEST;
}

interface IGetIngredientsSuccess {
    readonly type: typeof GET_INGREDIENTS_SUCCESS;
    items: IngredientType[];
}

interface IGetIngredientsFailed {
    readonly type: typeof GET_INGREDIENTS_FAILED;
}

export type TIngredientsActions = IGetIngredientsRequest | IGetIngredientsSuccess | IGetIngredientsFailed;

export const getIngredientsData = () => {
    return function (dispatch: Dispatch<TIngredientsActions | IError>) {
        dispatch({
            type: GET_INGREDIENTS_REQUEST
        });
        getIngredients()
            .then(res => {
                if (res && res.success) {
                    dispatch({
                        type: GET_INGREDIENTS_SUCCESS,
                        items: res.data
                    });
                } else {
                    dispatch({
                        type: GET_INGREDIENTS_FAILED
                    });
                }
            })
            .catch(err => {
                dispatch({
                    type: APPEND_ERROR_MODAL_TYPE,
                    message: err.message
                });
            });
    };
};
