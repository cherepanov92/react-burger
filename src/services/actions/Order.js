import { sendOrder } from '../../utils/api';
import { APPEND_ERROR_MODAL_TYPE, APPEND_MODAL_TYPE } from './Modal';

export const SENT_ORDER_REQUEST = 'SENT_ORDER_REQUEST';
export const SENT_ORDER_SUCCESS = 'SENT_ORDER_SUCCESS';
export const SENT_ORDER_FAILED = 'SENT_ORDER_FAILED';

export const sendOrderRequest = orderList => {
    return function (dispatch) {
        dispatch({
            type: SENT_ORDER_REQUEST
        });
        sendOrder(orderList)
            .then(res => {
                if (res && res.success) {
                    dispatch({
                        type: SENT_ORDER_SUCCESS,
                        name: res.name,
                        number: res.order.number
                    });
                    dispatch({ type: APPEND_MODAL_TYPE, modalType: 'order' });
                } else {
                    dispatch({
                        type: SENT_ORDER_FAILED
                    });
                }
            })
            .catch(err => {
                dispatch({
                    type: APPEND_ERROR_MODAL_TYPE,
                    message: err.message
                });
            });
    };
};
