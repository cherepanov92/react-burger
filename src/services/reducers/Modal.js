import { APPEND_ERROR_MODAL_TYPE, APPEND_MODAL_TYPE, REMOVE_MODAL_TYPE } from '../actions/Modal';

export default function ModalReducer(state = {}, action) {
    switch (action.type) {
        case APPEND_MODAL_TYPE: {
            return {
                ...state,
                modalType: action.modalType
            };
        }
        case APPEND_ERROR_MODAL_TYPE: {
            return {
                ...state,
                modalType: 'error',
                status: action.status
            };
        }
        case REMOVE_MODAL_TYPE: {
            return {
                ...state,
                modalType: null,
                status: null
            };
        }
        default: {
            return state;
        }
    }
}
