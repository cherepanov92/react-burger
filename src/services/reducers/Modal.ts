import { APPEND_ERROR_MODAL_TYPE, APPEND_MODAL_TYPE, REMOVE_MODAL_TYPE } from '../actions/Modal';
import { EnumModalType } from '../../utils/types';

type modalState = {
    modalType?: EnumModalType;
    status?: string;
    message?: string;
};

export default function modalReducer(state: modalState = {}, action: any) {
    switch (action.type) {
        case APPEND_MODAL_TYPE: {
            return {
                ...state,
                modalType: action.modalType
            };
        }
        case APPEND_ERROR_MODAL_TYPE: {
            return {
                ...state,
                modalType: EnumModalType.ERROR,
                status: action.status,
                message: action.message
            };
        }
        case REMOVE_MODAL_TYPE: {
            return {
                ...state,
                modalType: null,
                status: null
            };
        }
        default: {
            return state;
        }
    }
}
