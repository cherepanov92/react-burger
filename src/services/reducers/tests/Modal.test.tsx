import React from 'react';
import renderer from 'react-test-renderer';
import modalReducer, { InitialModalState as initialState } from '../Modal';
import { APPEND_ERROR_MODAL_TYPE, APPEND_MODAL_TYPE, REMOVE_MODAL_TYPE } from '../../actions/Modal';
import { EnumModalType } from '../../../utils/types';

describe('ModalReducer:', () => {
    it('Начальное состояние', () => {
        expect(modalReducer(initialState, {} as any)).toEqual(initialState);
        const tree = renderer.create(<>Начальное состояние</>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    describe('Добавление типа модального окна:', () => {
        it('заказ', () => {
            expect(
                modalReducer(initialState, {
                    type: APPEND_MODAL_TYPE,
                    modalType: EnumModalType.ORDER
                })
            ).toEqual({
                ...initialState,
                modalType: EnumModalType.ORDER
            });
        });

        it('ошибка', () => {
            expect(
                modalReducer(initialState, {
                    type: APPEND_MODAL_TYPE,
                    modalType: EnumModalType.ERROR
                })
            ).toEqual({
                ...initialState,
                modalType: EnumModalType.ERROR
            });
        });
    });

    describe('Добавление модального окна ошибок:', () => {
        it('все параметры', () => {
            expect(
                modalReducer(initialState, {
                    type: APPEND_ERROR_MODAL_TYPE,
                    status: '500',
                    message: 'Ошибка сервера'
                })
            ).toEqual({
                ...initialState,
                modalType: EnumModalType.ERROR,
                status: '500',
                message: 'Ошибка сервера'
            });
        });

        it('без статуса', () => {
            expect(
                modalReducer(initialState, {
                    type: APPEND_ERROR_MODAL_TYPE,
                    message: 'Ошибка сервера'
                })
            ).toEqual({
                ...initialState,
                modalType: EnumModalType.ERROR,
                status: null,
                message: 'Ошибка сервера'
            });
        });

        it('без сообщения', () => {
            expect(
                modalReducer(initialState, {
                    type: APPEND_ERROR_MODAL_TYPE,
                    status: '500',
                    message: null
                })
            ).toEqual({
                ...initialState,
                modalType: EnumModalType.ERROR,
                status: '500',
                message: null
            });
        });

        it('без статуса и сообщения', () => {
            expect(
                modalReducer(initialState, {
                    type: APPEND_ERROR_MODAL_TYPE,
                    status: null,
                    message: null
                })
            ).toEqual({
                ...initialState,
                modalType: EnumModalType.ERROR,
                status: null,
                message: null
            });
        });
    });

    it('Удаление данных о модальном окне', () => {
        expect(
            modalReducer(initialState, {
                type: REMOVE_MODAL_TYPE
            })
        ).toEqual({
            ...initialState,
            modalType: null,
            status: null
        });
    });
});
