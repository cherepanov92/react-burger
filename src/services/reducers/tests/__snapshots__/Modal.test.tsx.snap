// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalReducer: Начальное состояние 1`] = `"Начальное состояние"`;
